src/solc_0.7/common/Game/GameToken.sol

```diff
@@ -0,0 +1,320 @@
+        _minter = minter;
+        emit Minter(minter);
+    }
+
+    /// @notice Function to create a new GAME token
+    /// @param from The address of the one creating the game (may be different from msg.sender if metaTx)
+    /// @param to The address who will be assigned ownership of this game
+    /// @param assetIds The ids of the assets to add to this game
+    /// @param values the amount of each token id to add to game
+    /// @param editors The addresses to allow to edit (Can also be set later)
+    /// @return id The id of the new GAME token (erc721)
+    function createGame(
+        address from,
+        address to,
+        uint256[] memory assetIds,
+        uint256[] memory values,
+        address[] memory editors
+    ) public returns (uint256 id) {
+        // @review consider metaTransactions here. should we require "from", "to" or msg.sender to be the minter?
+        require(msg.sender == _minter || _minter == address(0), "INVALID_MINTER");
+        require(to != address(0), "DESTINATION_ZERO_ADDRESS");
+        require(to != address(this), "DESTINATION_GAME_CONTRACT");
+        require(assetIds.length == values.length, "INVALID_INPUT_LENGTHS");
+        uint256 gameId = _mintGame(from, to);
+
+        if (editors.length != 0) {
+            for (uint256 i = 0; i < editors.length; i++) {
+                _gameEditors[gameId][editors[i]] = true;
+            }
+        }
+        // a single asset is defined as 1 tokenID, and a value of 1 for that id. For anything else, use addMultipleAssets
+        if (assetIds.length != 0) {
+            if (assetIds.length == 1 && values[0] == 1) {
+                // Case: a single asset id with a value of 1
+                addSingleAsset(from, gameId, assetIds[0]);
+            } else {
+                // Case: Either multiple assetIds, or single assetId with value > 1
+                addMultipleAssets(from, gameId, assetIds, values);
+            }
+        }
+        emit AssetsAdded(gameId, assetIds, values);
+        return gameId;
+    }
+
+    /// @notice Function to add a single asset to an existing GAME
+    /// @param from The address of the one creating the game (may be different from msg.sender if metaTx)
+    /// @param gameId The id of the GAME to add asset to
+    /// @param assetId The id of the asset to add to GAME. Value is 1. If vaule needs to be > 1, use `addMultipleAssets(...)` instead.
+    function addSingleAsset(
+        address from,
+        uint256 gameId,
+        uint256 assetId
+    ) public {
+        require(msg.sender == _ownerOf(gameId) || _gameEditors[gameId][msg.sender], "ACCESS_DENIED");
+        // here "add" is from EnumerableSet.sol
+        _gameData[gameId]._assets.add(assetId);
+        uint256 assetValues = _gameData[gameId]._values[assetId];
+        // here "add" is from SafeMathWithRequire.sol
+        _gameData[gameId]._values[assetId] = assetValues.add(1);
+        _asset.transferFrom(from, address(this), assetId);
```

Comment:
Why not `safeTransferFrom` with empty data, as the `addMultipleAssets` function below uses `safeBatchTransferFrom`?

Reply:
Ha! indeed, this should use `safeTransferFrom`

Reply:
It should also be using the erc1155 version, not the erc721 one as the erc721 would fails on ingle item erc1155

Reply:
@wighawag I'm actually working on a refactor of that right now.